### DISCUSSION TITLE
Project Ideas from the community

### DISCUSSION THREAD

**Discussion by sloorush** [Dyalog/GSOC#6] [General]
Do you have an idea for a project that could work within Summer of Code, but is not on our list? Great! 💡 
This is where to put it, so we can comment and help you finalise it.

Please try to include as much as possible from this template:
 
```
Name of the project:
Estimated workload for you: 90 hours/ 175 hours /350 hours
Estimated difficulty for you: easy/medium/hard
Skills that will be needed:
    For example: HTML, CSS, JavaScript, (interest in) APL, Python, game programming, guide writing…
Description:
    Tell us your thoughts: what do you want to do, why and how?
Expected outcomes/benefits for Dyalog and the community
    What are the goals and how do you see us and the community benefitting from your idea?
Potential mentors/maintainers that would be relevant to talk to : 
```

**Comment by MaxCan-Code** [comment]
Examples: https://swift.org/gsoc2024
Guide from GSoC https://google.github.io/gsocguides/mentor/defining-a-project-ideas-list

**Comment by 11Kilobytes** [comment]
Name of the project: Adding trains support to Co-dfns
Estimated workload: 300 hours
Estimated difficulty for me: hard
Description: I want to go through the list of tacit functions from APLcart and make sure that all of them work with Co-dfns, by writing a test suite for each one of them. 
Expected outcomes/benefits for the dyalog community:
To start with, as a warmup, I've noticed that the `t0005` doesn't work (on my machine at least) so I want to see if I can fix that test and how long that would take me to see if this project is feasible. Then the idea is pretty simple: I'll record some videos where I pick a new train and try to get it working in `codfns.PS` and `codfns.TT`. The Dyalog community would obtain an APL compiler that would support a major feature of modern APL which would then vastly expand the set of existing programs which can be run on the GPU.
Potential mentors: @arcfide.

**Reply by sloorush** [reply]
Hi @11Kilobytes,
Thank you for your idea! Would you like to work on this idea as well?

@arcfide: Can you please take a look and give suggestions?

cc: @mkromberg @abrudz

**Reply by arcfide** [reply]
I think this is still at the high end of difficulty because introducing trains will be touching everything in the compiler. I'd put it at the max 350 hours instead of 300. 

My other comment is that trains support probably can't be handled realistically through "test driven" development directly. Test-case development will probably be sufficient to help work up a design for trains, but it is likely that you'll need to figure out a principled approach to parsing, compiling, and generating code for trains in order for this to work. You won't find it easier to go train by train, though the set of trains in APLCart is an excellent test suite. 

This would be quite a welcome addition, so I encourage this project to move forward.

**Reply by 11Kilobytes** [reply]
@arcfide, that's what I was thinking as well... I want to do this with the idea that I am always prototyping and can throw stuff away and start from scratch again as I learn. I like this project instead of the one about porting to a new runtime, since the only choices I would consider for a new runtime would be [Futhark](https://futhark-lang.org) and [Data.Array.Accelerate](https://www.acceleratehs.org/get-started.html) and only know how to do toy examples in each of those languages. 

@sloorush I would like to work on it, just seems kinda hard idk. I'll try the warmup tasks first and let y'all's know how I do.

**Reply by MaxCan-Code** [reply]
I'm just getting started with Co-dfns but I'd like to help too. The Nix config in Kabelo's fork[^0] helped me get set up quickly. I've built the runtime and running the test suite.

[^0]: https://github.com/11Kilobytes/Co-dfns

**Reply by arcfide** [reply]
> @arcfide, that's what I was thinking as well... I want to do this with the idea that I am always prototyping and can throw stuff away and start from scratch again as I learn.

I think that's the best way to think about it, honestly.
